OpenTrons/opentrons-api
# feat(robot-server): add instrument firmware update and monitoring capability

Closes RCORE-482, RCORE-483

# Overview

Adds endpoints to the robot-server to facilitate updating instruments' firmware. These endpoints will be primarily used to update instruments that get attached *after* a robot has finished auto-updating firmware of the instruments attached during a robot update.

The work for this feature started in a different branch whose code was reviewed [here](https://github.com/Opentrons/opentrons/pull/12209) if you want to take a look. Had to create a separate PR for untangling the multiple merge conflicts and for git tree readability.

# Test Plan

Test updating OT3 instruments and fetching the update status 
NOTE- you will need to temporarily modify the robot environment manually to make it so that it thinks there's an update available.-

1. Send a `POST` request to `/instruments/updates` endpoint with the mount specified in the body as
```
{"mount": <mount name>}
```
2. Fetch update status by sending `GET` request to `/instruments/updates/{:update_id}` where `update_id` is a unique ID generated by the `POST` update request.
3. Test that the following conditions make the update to fail with an appropriate error-
  - No instrument attached on the mount
  - Initiating an update while that mount is already updating
  - Initiating an update when there's no new firmware version available
  - Trying to fetch update status using a non-existent update ID
4. Check how long does it take for an update request to respond (useful for testing the timeout we can use for waiting for the update status)

Anything else you can think about.

# Changelog

- added the above two endpoints to update & fetch instruments.
- added an `UpdateProgressMonitor` class to keep track of the update processes that were started

# Review requests

- Please go through the code and check if it makes sense 
- Check that I have covered all test cases
- Check that inline comments/ todos make sense
- Test it on a robot! Bonus points if you could take screenshots of the responses so we can use them while wiring up the UI components.

# Risk assessment

Low for the code in edge. But could be detrimental to an instrument itself if an update process fails mid-way.
